### DISCUSSION TITLE
Add Date/TimeStamp suffix to build?

### DISCUSSION THREAD

**Discussion by jasonswearingen** [adamralph/minver#670] [Q&A] [answered]
Hello, I am new to Minver, it works great!

I am using it with VS2022.   I would like to add a timestamp suffix whenever the `Height` suffix is added.    So when I tag my latest commit `1.0.0` the Minver would also be `1.0.0`.   But when a commit is added, the next build would have a minVer of `1.0.0-alpha.0.20220206-1408`  or something similar.  (1408 being the current UTC time)  I'm not picky on the exact formatting, just would like to be able to add a timestamp to my nuget package build.

The purpose of all this is so that when I build a library project, another local project can detect a version change in the local nuget package, so I dont' need to commit when I'm doing local development  (fixing bugs in a library project used by another project)

Could someone please tell me how to do this?  Preferably without modifying environmental variables, as I think doing that requires some "external" workflow.  I'd like to be able to checkin whatever 

Thank you for your time!!!!!

**Comment by adamralph** [comment, marked as answer]
>  But when a commit is added...

If you just want the version number to change when a commit is added, then you don't need to do anything. That commit will increase the height by one.

**Reply by adamralph** [reply]
OK, I see that you later say:

> so I dont' need to commit when I'm doing local development

So I guess what you mean is:

> But after a commit is added, each successive build has a version of `1.0.1-alpha.0.yyymmdd-hhmm`.

(Note that those versions would be `1.0.1` pre-releases, since they follow the `1.0.0` tag.)

Regarding committing while doing local development: are you sure that's not the easiest option? I have a script in my PATH named `gw` which does `git commit -a -m "WIP"`, so each time I want the version to increment, I just type `gw`. Later, I squash those commits as required.

**Reply by adamralph** [reply]
Alternatively, you could do something like this, to add a timestamp to _all_ your pre-release builds:

```xml
<Target Name="MyTarget" AfterTargets="MinVer" Condition="'$(MinVerPreRelease)' != ''" >
  <PropertyGroup>
    <PackageVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)-$([System.DateTime]::Now.ToString(yyyyMMdd-hhmmss))</PackageVersion>
    <PackageVersion Condition="'$(MinVerBuildMetadata)' != ''">$(PackageVersion)+$(MinVerBuildMetadata)</PackageVersion>
    <Version>$(PackageVersion)</Version>
  </PropertyGroup>
</Target>
```

But note that this will also add a timestamp to your _tagged_ pre-release builds, not just those which have height. Currently there is no good way (as a consumer of MinVer) to tell if the current commit was tagged with a version or not. A variable could be added for that, but we should explore other ways to achieve what you want first.

**Reply by jasonswearingen** [reply]
> But after a commit is added, each successive build has a version of `1.0.1-alpha.0.yyyymmdd-hhmm`.

yep, that's what I'd like.

> Regarding committing while doing local development: are you sure that's not the easiest option? 

I think it would be better to not have to commit each fiddle with the code when debugging, only commit when I have the bug fixed and sorted out.    

I'm not sure if it's the new VS2022 preview or the Resharper extension, but with the local nuget package it's really easy for the consuming project to debug into the nuget package's source code, almost as "good" a debugging experience as if I just linked the library as a project dependency instead of a nuget package.   The only pain point is that I need a new version of the nuget package for the consuming project to update changes  (rebuilding the same version doesn't get picked up, even if I uninstall and reinstall the package.... the original version is cached somehow)

So right now every code change I want to try out, I need to commit to git and republish before using.    There is probably some easier way of doing rapid nuget debug/dev, but I'm pretty new to the ecosystem so I don't know any better :(

**Reply by jasonswearingen** [reply]
> Alternatively, you could do something like this, to add a timestamp to _all_ your pre-release builds:

Thank you, I pasted that snippet into my `.csproj` and it is generating what I want!

I think the condition you mentioned (will always emit timestamp if prerelease) is a fine compromise.  I can totally live with that :)

Thank you so much this is just about as perfect a solution as I could have hoped.

**Reply by adamralph** [reply]
I guess what you could also do right now, to avoid your official, tagged pre-releases having the timestamp suffix, is to suppress it on your build server. For example if using GitHub Actions:

```xml
<Target Name="MyTarget" AfterTargets="MinVer" Condition="'GITHUB_ACTIONS' != 'true' AND '$(MinVerPreRelease)' != ''" >
  <PropertyGroup>
    <PackageVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)-$([System.DateTime]::Now.ToString(yyyyMMdd-hhmmss))</PackageVersion>
    <PackageVersion Condition="'$(MinVerBuildMetadata)' != ''">$(PackageVersion)+$(MinVerBuildMetadata)</PackageVersion>
    <Version>$(PackageVersion)</Version>
  </PropertyGroup>
</Target>
```

**Reply by adamralph** [reply]
But more generally, I'm now considering adding a height variable, which would allow you do this:

```xml
<Target Name="MyTarget" AfterTargets="MinVer" Condition="'$(MinVerHeight)' != '0'" >
  <PropertyGroup>
    <PackageVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)-$([System.DateTime]::Now.ToString(yyyyMMdd-hhmmss))</PackageVersion>
    <PackageVersion Condition="'$(MinVerBuildMetadata)' != ''">$(PackageVersion)+$(MinVerBuildMetadata)</PackageVersion>
    <Version>$(PackageVersion)</Version>
  </PropertyGroup>
</Target>
```

**Reply by jasonswearingen** [reply]
I am pretty happy with this as it currently stands!  (I don't use github actions so can't comment on that, sadly).   But for my visual studio workflow this is pretty great, I only need to upgrade the nuget dependency in the consuming project when doing development.

> `Condition="'$(MinVerHeight)' != '0'"`

Seems like a useful additional variable :)

If I run into anything else I'll mention it here, but I'm pretty happy with this.
